daemons/based/based_transaction.c

```diff
@@ -0,0 +1,318 @@
+
+    if (transactions != NULL) {
+        g_hash_table_remove(transactions, client->id);
+    }
+}
+
+/*!
+ * \internal
+ * \brief Commit a given CIB client's transaction
+ *
+ * \param[in] client  Client whose transaction to commit
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+based_commit_transaction(const pcmk__client_t *client)
+{
+    GQueue *transaction = NULL;
+    int rc = pcmk_rc_ok;
+    xmlNodePtr saved_cib = the_cib;
+
+    CRM_ASSERT(client != NULL);
+
+    transaction = get_transaction(client);
+    if (transaction == NULL) {
+        return pcmk_rc_no_transaction;
+    }
+
+    // Apply all changes to a working copy of the CIB
+    the_cib = copy_xml(the_cib);
+
+    // Process all requests in the transaction
+    for (struct request_data *data = g_queue_pop_head(transaction);
+         data != NULL; data = g_queue_pop_head(transaction)) {
+
+        const char *op = crm_element_value(data->request, F_CIB_OPERATION);
+
+        rc = cib_process_request(data->request, data->privileged, data->client);
+        rc = pcmk_legacy2rc(rc);
+
+        if (rc != pcmk_rc_ok) {
+            crm_err("Aborting CIB transaction due to failed '%s' request: %s",
+                    op, pcmk_rc_str(rc));
+            crm_log_xml_err(data->request, "Failed request");
+            free_request_data(data);
+            free_xml(the_cib);
+            the_cib = saved_cib;
+            break;
+        }
+
+        crm_trace("Applied '%s' request to transaction working CIB", op);
+        crm_log_xml_trace(data->request, "Request");
+        free_request_data(data);
+    }
+
+    // Free the transaction and (if aborted) any remaining requests
+    based_discard_transaction(client);
+
+    // If successful, keep the working copy as the active CIB
+    if (rc == pcmk_rc_ok) {
+        crm_trace("CIB transaction complete");
+        free_xml(saved_cib);
+    }
+
+    /* @TODO:
+     * * Decide how to conclude, how to send a reply for the transaction commit,
+     *   and what the reply should include. A replace request seems redundant
+     *   since the active CIB is the working CIB and already contains all of our
+     *   changes, but it might be the simplest way. At least need to write to
+     *   disk.
+     * * Evaluate whether the current approach gives us the desired behavior for
+     *   diff/replace notifications. All relevant changes to the working CIB
+     *   trigger notifications, even if the transaction fails and the changes
+     *   are discarded. One possibility is to do a dry-run of all changes first
+     *   with cib_discard_reply and cib_inhibit_notify. That's not
+     *   straightforward though: if a request fails, we would want it to trigger
+     *   a callback and/or notification if applicable. Also cib_inhibit_notify
+     *   appears to affect only diff notifications, not replace notifications.
+     *   Perhaps do a dry-run of each request with cib_discard_reply and
+     *   cib_inhibit_notify.
```

Comment:
I still think a single notification at the end is more appropriate. The key feature of transactions is atomicity, so it makes sense to extend that to the notification. It also avoids the problem of failed transactions. The only other solution I can see is queuing the notifications instead of sending them, and sending them all if the transaction completes successfully.

Reply:
Single at the end is probably for the best for notifications. Honestly, notifications were an afterthought that came to me right before I submitted this.

My original reasoning for "respond to each request individually" was for callbacks. I was thinking it might be useful to have a callback for a particular request -- especially for a request that ends up failing.

The only *specific, current* example I had in mind was using a callback for adding a request to the controller's expected replacements table, so that the CIB replacement wouldn't trigger an election. However, if we're able to avoid notifications for individual requests in a transaction, then we can just add the commit request to the expected replacements table.

I'll think about how to suppress notifications (and probably callbacks, unless I think of a way they're useful) for individual requests. The new call option idea below was quarter-baked in a hurry. We'd only need this internally, so maybe a function argument, a new member in some struct, or an XML field. I'll take a look tonight.

Reply:
I'm hoping to use `result_cib` to replace the CIB. That feels cleaner than creating a new `PCMK__CIB_REQUEST_REPLACE` request. The only downside is that `based_commit_transaction()` doesn't **completely** commit the transaction. It commits all the requests to a working CIB, but the commit isn't really finished until `cib_process_command()` checks the return code and activates `result_cib`.

Can't properly test until I finish resolving the other known issues/to-dos.